include/crypto/ml_kem.h

```diff
@@ -0,0 +1,299 @@
+ *   defines the "noisy" linear equations: t = A * s + e.  Also the rank of
+ *   of the associated vectors.
+ * - "eta1" determines the amplitude of "s" and "e" vectors in key generation
+ *   and the "y" vector in ML-KEM encapsulation (K-PKE encryption).
+ * - "eta2" determines the amplitude of "e1" and "e2" noise terms in ML-KEM
+ *   encapsulation (K-PKE encryption).
+ * - "du" determines how many bits of each coefficient are retained in the
+ *   compressed form of the "u" vector in the encapsulation ciphertext.
+ * - "dv" determines how many bits of each coefficient are retained in the
+ *   compressed form of the "v" value in encapsulation ciphertext
+ * - "secbits" is required security strength of the RNG for the random inputs.
+ */
+
+/*
+ * The wire form of a losslessly encoded vector (12-bits per element)
+ */
+#  define ML_KEM_VECTOR_BYTES(rank) \
+    ((3 * ML_KEM_DEGREE / 2) * (rank))
+
+/*
+ * The wire-form public key consists of the lossless encoding of the vector
+ * "t" = "A" * "s" + "e", followed by public seed "rho".
+ */
+#  define ML_KEM_PUBKEY_BYTES(rank) \
+    (ML_KEM_VECTOR_BYTES(rank) + ML_KEM_RANDOM_BYTES)
+
+/*
+ * Our internal serialised private key concatenates serialisations of "s", the
+ * public key, the public key hash, and the failure secret "z".
+ */
+#  define ML_KEM_PRVKEY_BYTES(rank) \
+    (ML_KEM_VECTOR_BYTES(rank) + ML_KEM_PUBKEY_BYTES(rank) \
+     + ML_KEM_PKHASH_BYTES + ML_KEM_RANDOM_BYTES)
+
+/*
+ * Encapsulation produces a vector "u" and a scalar "v", whose coordinates
+ * (numbers modulo the ML-KEM prime "q") are lossily encoded using as "du" and
+ * "dv" bits, respectively.  This encoding is the ciphertext input for
+ * decapsulation.
+ */
+#  define ML_KEM_U_VECTOR_BYTES(rank, du) \
+    ((ML_KEM_DEGREE / 8) * (du) * (rank))
+#  define ML_KEM_V_SCALAR_BYTES(dv) \
+    ((ML_KEM_DEGREE / 8) * (dv))
+#  define ML_KEM_CTEXT_BYTES(rank, du, dv) \
+    (ML_KEM_U_VECTOR_BYTES(rank, du) + ML_KEM_V_SCALAR_BYTES(dv))
+
+/*
+ * Variant-specific constants and structures
+ * -----------------------------------------
+ */
+#  define ML_KEM_512_RANK       2
+#  define ML_KEM_512_ETA1       3
+#  define ML_KEM_512_ETA2       2
+#  define ML_KEM_512_DU         10
+#  define ML_KEM_512_DV         4
+#  define ML_KEM_512_RNGSEC     128
+
+#  define ML_KEM_768_RANK       3
+#  define ML_KEM_768_ETA1       2
+#  define ML_KEM_768_ETA2       2
+#  define ML_KEM_768_DU         10
+#  define ML_KEM_768_DV         4
+#  define ML_KEM_768_RNGSEC     192
+
+#  define ML_KEM_1024_RANK      4
+#  define ML_KEM_1024_ETA1      2
+#  define ML_KEM_1024_ETA2      2
+#  define ML_KEM_1024_DU        11
+#  define ML_KEM_1024_DV        5
+#  define ML_KEM_1024_RNGSEC    256
+
+/*
+ * External variant-specific API
+ * -----------------------------
+ */
+
+/*
+ * Each variant parameter set is associated with an ordinal number which
+ * represents that parameter set.
+ */
+# define ML_KEM_512     0
```

Comment:
Maybe rename to either MK_KEM_TYPE_512 or  MK_KEM_ID_512

Reply:
I'm passing on this one for now.